core/adapters/eth_tx.go

```diff
@@ -42,6 +44,93 @@ func (e *EthTx) TaskType() models.TaskType {
 // is not currently pending. Then it confirms the transaction was confirmed on
 // the blockchain.
 func (e *EthTx) Perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	if store.Config.EnableBulletproofTxManager() {
+		return e.perform(input, store)
+	} else {
+		return e.legacyPerform(input, store)
+	}
+}
+
+// TODO(sam): Move away from resuming this on every new head and do something more sensible
+func (e *EthTx) perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	trtx, err := store.FindEthTaskRunTxByTaskRunID(input.TaskRunID().UUID())
+	if err != nil {
+		err = errors.Wrap(err, "FindEthTaskRunTxByTaskRunID failed")
+		logger.Error(err)
+		return models.NewRunOutputError(err)
+	}
+	if trtx != nil {
+		return e.checkForConfirmation(*trtx, input, store)
+	}
+	return e.insertEthTx(input, store)
+}
+
+func (e *EthTx) checkForConfirmation(trtx models.EthTaskRunTx, input models.RunInput, store *store.Store) models.RunOutput {
+	switch trtx.EthTx.State {
+	case models.EthTxConfirmed:
+		receipt := models.EthReceipt{}
+		if err := store.GetRawDB().
```

Comment:
Why do we need to skip the advisory lock here?

Reply:
I'd flip it round... why would we need an advisory lock here?

Reply:
Yeah, I agree that we should start thinking of things in terms of "why lock" rather than "why not lock" if we want to get to our goal of horizontal scalability.  But at the same time, can you answer your own question for our (and posterity's) edification?

Reply:
OK, I'll try to explain. There is no need to take the Big Global Lock (tm) for this particular query because there aren't any concurrency issues.

It's a read only (no writes) and it doesn't matter when it happens. All state updates for `eth_txes` et al are atomic enforced via transaction so there is no danger of reading inconsistent data. If we read before the receipt gets written, we remain unconfirmed. If a receipt then gets written, we will confirm on the next head.

Reply:
As I wrote this out I suddenly realised we _can_ be in an inconsistent state, because we emit two queries non-transactionally:

1. Fetch ethtx
2. Switch on state
3. Fetch receipt

The state can be modified between 1 and 3 if the reorg detector marks it as unconfirmed.

Then in step 3 we would look for a receipt for an unconfirmed transaction. This probably wouldn't be the end of the world (we don't ever delete receipts) and the tx would simply be marked confirmed, but it isn't perfect behaviour. Will fix this.

Good catch @se3000.